concordium-node/src/grpc2.rs

```diff
@@ -2470,6 +2492,87 @@ pub mod server {
+                    })?;
+                    Some(permit)
+                }
+            };
+
+            let mut input = request.into_inner();
+            let energy_quota = self.dry_run_max_energy;
+            let deadline = Instant::now() + Duration::from_millis(self.dry_run_timeout_millis);
+            let mut dry_run = self.consensus.dry_run(energy_quota);
+            let output = async_stream::try_stream! {
+                let sleep = time::sleep_until(deadline);
+                tokio::pin!(sleep);
+                loop {
+                    let next = tokio::select! {
+                        request = input.next() => { Ok(request) }
+                        _ = &mut sleep => {
+                            Err(tonic::Status::deadline_exceeded("dry run deadline elapsed"))
+                        }
+                    }?;
+
+                    let Some(dry_run_request) = next else {
+                        // The stream of requests has been closed by the sender.
+                        break
+                    };
+                    let request = dry_run_request
+                        .map_err(|_| tonic::Status::invalid_argument("invalid dry run request"))?;
+                    use crate::grpc2::types::dry_run_request::Request::*;
+                    match request.request.require()? {
+                        LoadBlockState(block_hash_input) => {
+                            yield dry_run.load_block_state(&block_hash_input)?
+                        }
+                        StateQuery(query) => {
+                            use crate::grpc2::types::dry_run_state_query::Query::*;
+                            match query.query.require()? {
+                                GetAccountInfo(account) => yield dry_run.get_account_info(&account)?,
+                                GetInstanceInfo(instance) => {
+                                    yield dry_run.get_instance_info(&instance)?
+                                }
+                                InvokeInstance(invoke_instance_input) => {
+                                    yield dry_run.invoke_instance(&invoke_instance_input)?
+                                }
+                            }
+                        }
+                        StateOperation(operation) => {
+                            use crate::grpc2::types::dry_run_state_operation::Operation::*;
+                            match operation.operation.require()? {
+                                SetTimestamp(timestamp) => yield dry_run.set_timestamp(timestamp)?,
+                                MintToAccount(mint) => yield dry_run.mint_to_account(mint)?,
+                                RunTransaction(run_transaction_input) => {
+                                    yield dry_run.transaction(run_transaction_input)?
+                                }
+                            }
+                        }
+                    }
```

Comment:
Would it not make more sense to make a separate stream object which owns the necessary data instead of bringing in this dependency and using rather obscure methods.

Something like

```rust
struct Response {
    pub dry_run: DryRun,
    pub input:   tonic::Streaming<self::types::DryRunRequest>,
    pub timeout: std::pin::Pin<Box<tokio::time::Sleep>>,
    pub permit: Option<OwnedSemaphorePermit>,
    pub done: bool,
}

impl futures::Stream for Response {
    type Item = tonic::Result<Vec<u8>>;

    fn poll_next(
        mut self: std::pin::Pin<&mut Self>,
        cx: &mut std::task::Context<'_>,
    ) -> std::task::Poll<Option<Self::Item>> {
        use std::task::Poll::Ready;
        if self.done {
            return Ready(None);
        };
        if let std::task::Poll::Ready(_) = std::pin::Pin::new(&mut self.timeout).poll(cx) {
            self.done = true;
            return Ready(Some(Err(tonic::Status::deadline_exceeded("..."))));
        };
        let Some(dry_run_request) = futures::ready!(std::pin::Pin::new(&mut self.input).poll_next(cx)) else {
            return Ready(None);
        };
        let request = dry_run_request
            .map_err(|_| tonic::Status::invalid_argument("invalid dry run request"))?;
        use crate::grpc2::types::dry_run_request::Request::*;
        let result = match request.request.require()? {
            LoadBlockState(block_hash_input) => self.dry_run.load_block_state(&block_hash_input),
            StateQuery(query) => {
                use crate::grpc2::types::dry_run_state_query::Query::*;
                match query.query.require()? {
                    GetAccountInfo(account) => self.dry_run.get_account_info(&account),
                    GetInstanceInfo(instance) => self.dry_run.get_instance_info(&instance),
                    InvokeInstance(invoke_instance_input) => {
                        self.dry_run.invoke_instance(&invoke_instance_input)
                    }
                }
            }
            StateOperation(operation) => {
                use crate::grpc2::types::dry_run_state_operation::Operation::*;
                match operation.operation.require()? {
                    SetTimestamp(timestamp) => self.dry_run.set_timestamp(timestamp),
                    MintToAccount(mint) => self.dry_run.mint_to_account(mint),
                    RunTransaction(run_transaction_input) => {
                        self.dry_run.transaction(run_transaction_input)
                    }
                }
            }
        };
        Ready(Some(result))
    }
}
```